larq_compute_engine/core/packbits.h

```diff
@@ -382,19 +353,30 @@ inline void packbits_array(const TIn* input_array, const std::size_t n,
     memcpy(padding_buffer.data(), in, elements_left * sizeof(TIn));
     for (size_t i = elements_left; i < bitwidth; ++i)
       padding_buffer[i] = zero_point;
-    pack_bitfield<bitpack_order>(padding_buffer.data(), out, zero_point);
+    pack_bitfield(padding_buffer.data(), out, zero_point);
   }
 }
 
 // Bitpacks each row of a row-major matrix
-template <BitpackOrder bitpack_order, class TIn, class TOut_>
+template <class TIn, class TOut_>
 inline void packbits_matrix(const TIn* input, const std::size_t input_num_rows,
                             const std::size_t input_num_cols, TOut_* output,
                             const std::int32_t zero_point = 0) {
   // Force the types to be unsigned so that the function can be called on signed
   // types as well
   using TOut = typename std::make_unsigned<TOut_>::type;
 
+#ifdef __aarch64__
+  if (!CE_IS_BIG_ENDIAN && std::is_same<TIn, float>::value &&
+      std::is_same<TOut, std::uint64_t>::value && input_num_cols % 64 == 0 &&
+      zero_point == 0) {
+    packbits_aarch64_64(reinterpret_cast<const float*>(input),
+                        input_num_rows * (input_num_cols / 64),
+                        reinterpret_cast<std::uint64_t*>(output));
+    return;
+  }
+#endif
```

Comment:
One key difference is that previously the optimised assembly bitpacking code bitpacked 64 floats into a single uint64, and was invoked multiple times from nested C++ loops. Instead, in this PR the optimised assembly bitpacking code bitpacks a (long, multiple of 64) sequence of floats into a sequence of uint64s. Doing this makes it easier to schedule the memory accesses for the next block (loads into registers) optimally alongside the computation for the current block.

However, there is technically a regression here in that previously an attempt to bitpack a tensor of 129 channels would use optimised assembly for the first 128 channels, and then slower reference code for the single remaining channel. In this PR, such a tensor would be bitpacked only by reference code. I'm not sure how much of a problem that is - most models use multiples of 64 input channels pretty much always (e.g. all the QuickNet family of models).

Reply:
How about we do this optimization more generally, regardless of aarch64: if `num_cols % bitwidth == 0` then call the packbits function with just one set of blocks. Else, call the packbits function in a loop. Then the aarch64 optimized code would be used for both.
(I had wanted to do that long ago: https://github.com/larq/compute-engine/issues/124 )